File: smart-account-demo/shared/src/commonMain/kotlin/com/soneso/smartdemo/ui/screens/WalletCreationScreen.kt
package com.soneso.smartdemo.ui.screens

/**
 * Wallet creation screen: collects a username, triggers wallet creation via WalletCreationFlow,
 * and displays the resulting credential, contract address, and initial balances.
 *
 * When autoSubmit=false, the passkey is registered but the contract is not deployed on-chain.
 * In this case the screen shows an info card with a "Deploy Now" button instead of the full
 * success card with balances.
 */

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.material3.Checkbox
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import cafe.adriel.voyager.core.screen.Screen
import cafe.adriel.voyager.navigator.LocalNavigator
import cafe.adriel.voyager.navigator.currentOrThrow
import com.soneso.smartdemo.flows.WalletCreationResult
import com.soneso.smartdemo.flows.createWallet
import com.soneso.smartdemo.flows.deployPendingAndProvision
import com.soneso.smartdemo.platform.Clipboard
import com.soneso.smartdemo.platform.getClipboard
import com.soneso.smartdemo.state.ActivityLogState
import com.soneso.smartdemo.state.DemoState
import com.soneso.smartdemo.util.isUserCancellation
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

class WalletCreationScreen : Screen {
    @OptIn(ExperimentalMaterial3Api::class)
    @Composable
    override fun Content() {
        val navigator = LocalNavigator.currentOrThrow
        val scope = rememberCoroutineScope()
        val snackbarHostState = remember { SnackbarHostState() }
        val clipboard = remember { getClipboard() }

        // Form input state
        var username by remember { mutableStateOf("") }
        var autoSubmit by remember { mutableStateOf(true) }

        // Loading / result state
        var isLoading by remember { mutableStateOf(false) }
        var progressMessage by remember { mutableStateOf("") }
        var errorMessage by remember { mutableStateOf<String?>(null) }
        var infoMessage by remember { mutableStateOf<String?>(null) }
        var createResult by remember { mutableStateOf<WalletCreationResult?>(null) }

        // Deploy-from-creation-screen state (for autoSubmit=false results)
        var isDeploying by remember { mutableStateOf(false) }
        var deployErrorMessage by remember { mutableStateOf<String?>(null) }
        var deployedSuccessfully by remember { mutableStateOf(false) }

        Scaffold(
            snackbarHost = { SnackbarHost(snackbarHostState) },
            topBar = {
                TopAppBar(
                    title = { Text("Create Wallet") },
                    navigationIcon = {
                        IconButton(onClick = { navigator.pop() }) {
                            Icon(Icons.AutoMirrored.Filled.ArrowBack, "Back")
                        }
                    },
                    colors = TopAppBarDefaults.topAppBarColors(
                        containerColor = MaterialTheme.colorScheme.primary,
                        titleContentColor = MaterialTheme.colorScheme.onPrimary,
                        navigationIconContentColor = MaterialTheme.colorScheme.onPrimary
                    )
                )
            }
        ) { padding ->
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding)
                    .padding(16.dp)
                    .verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                // Info Card
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.surfaceVariant
                    )
                ) {
                    Column(
                        modifier = Modifier.padding(16.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        Text(
                            text = "Wallet Creation",
                            style = MaterialTheme.typography.titleMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                        Text(
                            text = "Creating a wallet will register a passkey with your device and deploy a smart account contract to the Stellar network. The passkey is used to authenticate and sign transactions.",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }

                // Username Input
                OutlinedTextField(
                    value = username,
                    onValueChange = { username = it },
                    label = { Text("Passkey Name") },
                    modifier = Modifier.fillMaxWidth(),
                    enabled = !isLoading && createResult == null,
                    singleLine = true
                )

                // Error Message
                if (errorMessage != null) {
                    Card(
                        modifier = Modifier.fillMaxWidth(),
                        colors = CardDefaults.cardColors(
                            containerColor = MaterialTheme.colorScheme.errorContainer
                        )
                    ) {
                        SelectionContainer {
                            Text(
                                text = errorMessage!!,
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onErrorContainer,
                                modifier = Modifier.padding(16.dp)
                            )
                        }
                    }
                }

                // Info Message
                if (infoMessage != null) {
                    Card(
                        modifier = Modifier.fillMaxWidth(),
                        colors = CardDefaults.cardColors(
                            containerColor = MaterialTheme.colorScheme.secondaryContainer
                        )
                    ) {
                        Text(
                            text = infoMessage!!,
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSecondaryContainer,
                            modifier = Modifier.padding(16.dp)
                        )
                    }
                }

                // Progress indicator (shown during wallet creation)
                if (isLoading) {
                    Card(
                        modifier = Modifier.fillMaxWidth(),
                        colors = CardDefaults.cardColors(
                            containerColor = MaterialTheme.colorScheme.surfaceVariant
                        )
                    ) {
                        Column(
                            modifier = Modifier.fillMaxWidth().padding(16.dp),
                            horizontalAlignment = Alignment.CenterHorizontally,
                            verticalArrangement = Arrangement.spacedBy(12.dp)
                        ) {
                            CircularProgressIndicator(
                                modifier = Modifier.size(32.dp),
                                strokeWidth = 3.dp
                            )
                            Text(
                                text = progressMessage.ifEmpty { "Creating..." },
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }
                }

                // Auto-submit checkbox
                if (createResult == null && !isLoading) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Checkbox(
                            checked = autoSubmit,
                            onCheckedChange = { autoSubmit = it }
                        )
                        Text(
                            text = "Auto-deploy after passkey registration",
                            style = MaterialTheme.typography.bodyMedium,
                            modifier = Modifier.clickable { autoSubmit = !autoSubmit }
                        )
                    }
                }

                // Create Wallet Button
                if (createResult == null && !isLoading) {
                    Button(
                        onClick = {
                            scope.launch {
                                isLoading = true
                                errorMessage = null
                                infoMessage = null

                                try {
                                    val result = createWallet(username, autoSubmit) { progressMessage = it }
                                    createResult = result
                                    // If deployed, mark as such for the result card
                                    deployedSuccessfully = DemoState.isDeployed
                                } catch (e: Throwable) {
                                    val message = e.message ?: "Unknown error"
                                    if (isUserCancellation(message)) {
                                        infoMessage = "Passkey registration cancelled by user"
                                        ActivityLogState.info("User cancelled passkey registration")
                                    } else {
                                        errorMessage = "Failed to create wallet: $message\n\n" +
                                            "If a passkey was registered before the failure, " +
                                            "go to Connect Wallet and check Pending Deployments " +
                                            "to retry the deployment."
                                        ActivityLogState.error(message)
                                    }
                                } finally {
                                    isLoading = false
                                }
                            }
                        },
                        modifier = Modifier.fillMaxWidth(),
                        enabled = username.isNotBlank() && DemoState.kit != null
                    ) {
                        Text("Create Wallet")
                    }
                }

                // Result Section
                if (createResult != null) {
                    val result = createResult!!

                    if (deployedSuccessfully) {
                        // Deployed: show full success card with balances
                        DeployedResultCard(
                            result = result,
                            clipboard = clipboard,
                            snackbarHostState = snackbarHostState,
                            scope = scope
                        )
                    } else {
                        // Not deployed: show passkey-registered card with deploy option
                        UndeployedResultCard(
                            result = result,
                            clipboard = clipboard,
                            snackbarHostState = snackbarHostState,
                            scope = scope,
                            isDeploying = isDeploying,
                            deployErrorMessage = deployErrorMessage,
                            onDeploy = {
                                scope.launch {
                                    isDeploying = true
                                    deployErrorMessage = null
                                    try {
                                        val provisionResult = deployPendingAndProvision(
                                            credentialId = result.credentialId,
                                            onProgress = { progressMessage = it }
                                        )
                                        if (provisionResult.success) {
                                            deployedSuccessfully = true
                                            createResult = result.copy(
                                                xlmBalance = provisionResult.xlmBalance,
                                                demoTokenBalance = provisionResult.demoTokenBalance
                                            )
                                        } else {
                                            deployErrorMessage = "Deployment failed: ${provisionResult.error}"
                                            ActivityLogState.error("Deployment failed: ${provisionResult.error}")
                                        }
                                    } catch (e: Throwable) {
                                        deployErrorMessage = "Deployment failed: ${e.message}"
                                        ActivityLogState.error("Deployment failed: ${e.message}")
                                    } finally {
                                        isDeploying = false
                                    }
                                }
                            }
                        )
                    }

                    // Go to Main Screen Button (disabled during deployment)
                    Button(
                        onClick = { navigator.pop() },
                        modifier = Modifier.fillMaxWidth(),
                        enabled = !isDeploying
                    ) {
                        Text("Go to Main Screen")
                    }
                }

                Spacer(modifier = Modifier.height(16.dp))
            }
        }
    }

    @Composable
    private fun DeployedResultCard(
        result: WalletCreationResult,
        clipboard: Clipboard,
        snackbarHostState: SnackbarHostState,
        scope: CoroutineScope
    ) {
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.primaryContainer
            )
        ) {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Text(
                    text = "Wallet Created Successfully",
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onPrimaryContainer
                )

                ResultField(
                    label = "Credential ID",
                    value = result.credentialId,
                    clipboard = clipboard,
                    snackbarHostState = snackbarHostState,
                    scope = scope
                )

                ResultField(
                    label = "Contract Address",
                    value = result.contractId,
                    clipboard = clipboard,
                    snackbarHostState = snackbarHostState,
                    scope = scope
                )

                if (result.transactionHash != null) {
                    ResultField(
                        label = "Transaction Hash",
                        value = result.transactionHash,
                        clipboard = clipboard,
                        snackbarHostState = snackbarHostState,
                        scope = scope
                    )
                }

                Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                    Text(
                        text = "Balance",
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.primary
                    )
                    Text(
                        text = "${result.xlmBalance} XLM",
                        style = MaterialTheme.typography.bodyMedium,
                        fontFamily = FontFamily.Monospace,
                        color = MaterialTheme.colorScheme.onPrimaryContainer
                    )
                    if (result.demoTokenBalance != null) {
                        Text(
                            text = "${result.demoTokenBalance} DEMO",
                            style = MaterialTheme.typography.bodyMedium,
                            fontFamily = FontFamily.Monospace,
                            color = MaterialTheme.colorScheme.onPrimaryContainer
                        )
                    }
                }
            }
        }
    }

    @Composable
    private fun UndeployedResultCard(
        result: WalletCreationResult,
        clipboard: Clipboard,
        snackbarHostState: SnackbarHostState,
        scope: CoroutineScope,
        isDeploying: Boolean,
        deployErrorMessage: String?,
        onDeploy: () -> Unit
    ) {
        // Warning amber color for undeployed state
        val warningContainerColor = Color(0xFFFFF3E0) // Orange 50
        val warningTextColor = Color(0xFFE65100) // Orange 900

        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.secondaryContainer
            )
        ) {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Text(
                    text = "Passkey Registered",
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSecondaryContainer
                )

                ResultField(
                    label = "Credential ID",
                    value = result.credentialId,
                    clipboard = clipboard,
                    snackbarHostState = snackbarHostState,
                    scope = scope,
                    textColor = MaterialTheme.colorScheme.onSecondaryContainer
                )

                ResultField(
                    label = "Contract Address (derived)",
                    value = result.contractId,
                    clipboard = clipboard,
                    snackbarHostState = snackbarHostState,
                    scope = scope,
                    textColor = MaterialTheme.colorScheme.onSecondaryContainer
                )

                // Warning message
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(
                        containerColor = warningContainerColor
                    )
                ) {
                    Text(
                        text = "The wallet contract has not been deployed to the network yet. Deploy it now or later from the Connect Wallet screen.",
                        style = MaterialTheme.typography.bodyMedium,
                        color = warningTextColor,
                        modifier = Modifier.padding(12.dp)
                    )
                }

                // Deploy error
                if (deployErrorMessage != null) {
                    Card(
                        modifier = Modifier.fillMaxWidth(),
                        colors = CardDefaults.cardColors(
                            containerColor = MaterialTheme.colorScheme.errorContainer
                        )
                    ) {
                        SelectionContainer {
                            Text(
                                text = deployErrorMessage,
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onErrorContainer,
                                modifier = Modifier.padding(12.dp)
                            )
                        }
                    }
                }

                // Deploy progress
                if (isDeploying) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.Center,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(24.dp),
                            strokeWidth = 2.dp
                        )
                        Spacer(modifier = Modifier.padding(horizontal = 8.dp))
                        Text(
                            text = "Deploying contract...",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSecondaryContainer
                        )
                    }
                }

                // Deploy Now button
                Button(
                    onClick = onDeploy,
                    modifier = Modifier.fillMaxWidth(),
                    enabled = !isDeploying
                ) {
                    Text(if (isDeploying) "Deploying..." else "Deploy Now")
                }
            }
        }
    }

    @Composable
    private fun ResultField(
        label: String,
        value: String,
        clipboard: Clipboard,
        snackbarHostState: SnackbarHostState,
        scope: CoroutineScope,
        textColor: Color = MaterialTheme.colorScheme.onPrimaryContainer
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            Text(
                text = label,
                style = MaterialTheme.typography.labelMedium,
                color = textColor
            )
            Text(
                text = value,
                style = MaterialTheme.typography.bodyMedium,
                fontFamily = FontFamily.Monospace,
                color = textColor,
                modifier = Modifier.clickable {
                    scope.launch {
                        clipboard.copyToClipboard(value)
                        snackbarHostState.showSnackbar("$label copied to clipboard")
                    }
                }
            )
            Text(
                text = "Tap to copy",
                style = MaterialTheme.typography.labelSmall,
                color = textColor.copy(alpha = 0.6f)
            )
        }
    }
}

File: smart-account-demo/macosApp/StellarSmartDemo/Views/WalletCreationScreen.swift
//
//  WalletCreationScreen.swift
//  StellarSmartDemo
//
//  Copyright (c) 2026 Soneso. All rights reserved.
//

import SwiftUI
import shared

/// Screen that collects a display name, triggers passkey registration, and optionally deploys a
/// smart account contract to the Stellar testnet.
///
/// The screen mirrors the Compose `WalletCreationScreen` and covers the full wallet creation
/// lifecycle:
/// 1. The user enters a passkey display name.
/// 2. Tapping "Create Wallet" calls `MacOSBridge.createWallet(username:autoSubmit:onProgress:)`.
/// 3. Progress messages are shown in a live card while the operation is in flight.
/// 4. On success:
///    - If autoSubmit=true: a green result card shows credentials, balances, and "Go to Main Screen".
///    - If autoSubmit=false: a blue info card shows the passkey was registered but not deployed,
///      with a "Deploy Now" button to retry deployment inline.
/// 5. User cancellation (Touch ID dismissed) is distinguished from genuine errors.
struct WalletCreationScreen: View {

    // MARK: - Environment

    @EnvironmentObject private var bridgeWrapper: MacOSBridgeWrapper
    @EnvironmentObject private var appState: AppState
    @ObservedObject var toastManager: ToastManager
    @Environment(\.dismiss) private var dismiss

    // MARK: - State

    @State private var username: String = ""
    @State private var autoSubmit: Bool = true
    @State private var isLoading: Bool = false
    @State private var progressMessage: String = ""
    @State private var errorMessage: String? = nil
    @State private var infoMessage: String? = nil
    @State private var createResult: WalletCreationResult? = nil

    // Deploy-from-creation-screen state (for autoSubmit=false results)
    @State private var isDeploying: Bool = false
    @State private var deployErrorMessage: String? = nil
    @State private var deployedSuccessfully: Bool = false

    // MARK: - Derived State

    private var isButtonEnabled: Bool {
        !username.trimmingCharacters(in: .whitespaces).isEmpty
            && bridgeWrapper.isKitInitialized
            && !isLoading
            && createResult == nil
    }

    // MARK: - Body

    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                descriptionCard
                usernameField
                if let error = errorMessage { ErrorCard(message: error) }
                if let info = infoMessage { infoCard(message: info) }
                if isLoading { progressCard }
                if createResult == nil && !isLoading { autoSubmitToggle }
                if createResult == nil && !isLoading { createButton }
                if let result = createResult { resultSection(result: result) }
                Spacer().frame(height: 16)
            }
            .padding(16)
        }
        .background(Material3Colors.background)
        .navigationToolbar(title: "Create Wallet")
    }

    // MARK: - Description Card

    private var descriptionCard: some View {
        InfoCard(title: "Wallet Creation", color: .variant) {
            Text(
                "Creating a wallet will register a passkey with your device and deploy a smart " +
                "account contract to the Stellar network. The passkey is used to authenticate " +
                "and sign transactions."
            )
            .font(.system(size: 13))
            .foregroundStyle(Material3Colors.onSurfaceVariant)
        }
    }

    // MARK: - Username Input

    private var usernameField: some View {
        ValidationTextField(
            label: "Passkey Name",
            text: $username,
            placeholder: "Enter a display name for your passkey",
            isEnabled: !isLoading && createResult == nil
        )
    }

    // MARK: - Info Card (user cancellation)

    @ViewBuilder
    private func infoCard(message: String) -> some View {
        Text(message)
            .font(.system(size: 13))
            .foregroundStyle(Material3Colors.onSecondaryContainer)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(16)
            .background(Material3Colors.secondaryContainer)
            .cornerRadius(8)
    }

    // MARK: - Progress Card

    private var progressCard: some View {
        VStack(spacing: 12) {
            ProgressView()
                .controlSize(.large)
                .scaleEffect(1.1)
                .frame(width: 32, height: 32)
            Text(progressMessage.isEmpty ? "Creating..." : progressMessage)
                .font(.system(size: 13))
                .foregroundStyle(Material3Colors.onSurfaceVariant)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(16)
        .background(Material3Colors.surfaceVariant)
        .cornerRadius(8)
    }

    // MARK: - Auto-submit Toggle

    private var autoSubmitToggle: some View {
        Toggle("Auto-deploy after passkey registration", isOn: $autoSubmit)
            .toggleStyle(.checkbox)
            .font(.system(size: 13))
            .foregroundStyle(Material3Colors.onSurface)
    }

    // MARK: - Create Button

    private var createButton: some View {
        LoadingButton(
            action: startWalletCreation,
            isLoading: isLoading,
            isEnabled: isButtonEnabled,
            text: "Create Wallet",
            loadingText: "Creating wallet...",
            style: .filled
        )
    }

    // MARK: - Result Section

    @ViewBuilder
    private func resultSection(result: WalletCreationResult) -> some View {
        VStack(spacing: 16) {
            if deployedSuccessfully {
                deployedCard(result: result)
            } else {
                undeployedCard(result: result)
            }

            LoadingButton(
                action: { dismiss() },
                isLoading: false,
                isEnabled: !isDeploying,
                text: "Go to Main Screen",
                loadingText: "",
                style: .filled
            )
        }
    }

    // MARK: - Deployed Success Card

    @ViewBuilder
    private func deployedCard(result: WalletCreationResult) -> some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Wallet Created Successfully")
                .font(.headline)
                .fontWeight(.semibold)
                .foregroundStyle(Material3Colors.onPrimaryContainer)

            CopyableField(
                label: "Credential ID",
                value: result.credentialId,
                textColor: Material3Colors.onPrimaryContainer,
                labelColor: Material3Colors.onPrimaryContainer,
                monospace: true,
                onCopy: { toastManager.show("Credential ID copied to clipboard") }
            )

            CopyableField(
                label: "Contract Address",
                value: result.contractId,
                textColor: Material3Colors.onPrimaryContainer,
                labelColor: Material3Colors.onPrimaryContainer,
                monospace: true,
                onCopy: { toastManager.show("Contract address copied to clipboard") }
            )

            if let txHash = result.transactionHash {
                CopyableField(
                    label: "Transaction Hash",
                    value: txHash,
                    textColor: Material3Colors.onPrimaryContainer,
                    labelColor: Material3Colors.onPrimaryContainer,
                    monospace: true,
                    onCopy: { toastManager.show("Transaction hash copied to clipboard") }
                )
            }

            balanceSection(result: result)
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Material3Colors.primaryContainer)
        .cornerRadius(8)
    }

    // MARK: - Undeployed Card

    @ViewBuilder
    private func undeployedCard(result: WalletCreationResult) -> some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Passkey Registered")
                .font(.headline)
                .fontWeight(.semibold)
                .foregroundStyle(Material3Colors.onSecondaryContainer)

            CopyableField(
                label: "Credential ID",
                value: result.credentialId,
                textColor: Material3Colors.onSecondaryContainer,
                labelColor: Material3Colors.onSecondaryContainer,
                monospace: true,
                onCopy: { toastManager.show("Credential ID copied to clipboard") }
            )

            CopyableField(
                label: "Contract Address (derived)",
                value: result.contractId,
                textColor: Material3Colors.onSecondaryContainer,
                labelColor: Material3Colors.onSecondaryContainer,
                monospace: true,
                onCopy: { toastManager.show("Contract address copied to clipboard") }
            )

            // Warning message
            warningBanner(
                text: "The wallet contract has not been deployed to the network yet. " +
                      "Deploy it now or later from the Connect Wallet screen."
            )

            // Deploy error
            if let error = deployErrorMessage {
                Text(error)
                    .font(.system(size: 13))
                    .foregroundStyle(Material3Colors.onErrorContainer)
                    .textSelection(.enabled)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(12)
                    .background(Material3Colors.errorContainer)
                    .cornerRadius(6)
            }

            // Deploy progress
            if isDeploying {
                HStack(spacing: 8) {
                    ProgressView()
                        .controlSize(.small)
                    Text("Deploying contract...")
                        .font(.system(size: 13))
                        .foregroundStyle(Material3Colors.onSecondaryContainer)
                }
                .frame(maxWidth: .infinity)
            }

            // Deploy Now button
            LoadingButton(
                action: deployFromCreationScreen,
                isLoading: isDeploying,
                isEnabled: !isDeploying,
                text: "Deploy Now",
                loadingText: "Deploying...",
                style: .filled
            )
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Material3Colors.secondaryContainer)
        .cornerRadius(8)
    }

    @ViewBuilder
    private func balanceSection(result: WalletCreationResult) -> some View {
        BalanceRows(
            label: "Balance",
            labelColor: Material3Colors.primary,
            values: balanceValues(result: result),
            valueFont: .system(.callout, design: .monospaced),
            valueColor: Material3Colors.onPrimaryContainer
        )
    }

    /// Formatted balance lines: XLM always, DEMO only when minted during provisioning.
    private func balanceValues(result: WalletCreationResult) -> [String] {
        var values = ["\(result.xlmBalance) XLM"]
        if let demoBalance = result.demoTokenBalance {
            values.append("\(demoBalance) DEMO")
        }
        return values
    }

    @ViewBuilder
    private func warningBanner(text: String) -> some View {
        Text(text)
            .font(.system(size: 13))
            .foregroundStyle(Material3Colors.warningText)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(12)
            .background(Material3Colors.warningBackground)
            .cornerRadius(6)
    }

    // MARK: - Actions

    private func startWalletCreation() {
        isLoading = true
        errorMessage = nil
        infoMessage = nil

        Task {
            do {
                let result = try await bridgeWrapper.bridge.createWallet(
                    username: username,
                    autoSubmit: autoSubmit,
                    onProgress: { progress in
                        DispatchQueue.main.async {
                            self.progressMessage = progress
                        }
                    }
                )
                await MainActor.run {
                    self.createResult = result
                    self.isLoading = false
                    appState.sync(from: bridgeWrapper.bridge)
                    self.deployedSuccessfully = appState.isDeployed
                    if self.deployedSuccessfully {
                        toastManager.show("Wallet created successfully")
                    } else {
                        toastManager.show("Passkey registered (deployment pending)")
                    }
                }
            } catch {
                await MainActor.run {
                    let message = error.localizedDescription
                    if bridgeWrapper.bridge.isUserCancellation(message: message) {
                        self.infoMessage = "Passkey registration cancelled by user"
                        ActivityLogState.shared.info(message: "User cancelled passkey registration")
                    } else {
                        self.errorMessage =
                            "Failed to create wallet: \(message)\n\n" +
                            "If a passkey was registered before the failure, " +
                            "go to Connect Wallet and check Pending Deployments " +
                            "to retry the deployment."
                        ActivityLogState.shared.error(message: message)
                    }
                    appState.syncActivityLog(from: bridgeWrapper.bridge)
                    self.isLoading = false
                }
            }
        }
    }

    private func deployFromCreationScreen() {
        guard let result = createResult else { return }
        isDeploying = true
        deployErrorMessage = nil

        Task {
            do {
                let provisionResult = try await bridgeWrapper.bridge.deployPendingAndProvision(
                    credentialId: result.credentialId,
                    onProgress: { progress in
                        DispatchQueue.main.async {
                            self.progressMessage = progress
                        }
                    }
                )
                await MainActor.run {
                    appState.sync(from: bridgeWrapper.bridge)
                    if provisionResult.success {
                        self.deployedSuccessfully = true
                        // Update the result with balances from provisioning
                        self.createResult = WalletCreationResult(
                            credentialId: result.credentialId,
                            contractId: result.contractId,
                            transactionHash: result.transactionHash,
                            nickname: result.nickname,
                            xlmBalance: provisionResult.xlmBalance,
                            demoTokenBalance: provisionResult.demoTokenBalance
                        )
                        toastManager.show("Wallet deployed successfully")
                    } else {
                        self.deployErrorMessage = "Deployment failed: \(provisionResult.error ?? "Unknown error")"
                    }
                    self.isDeploying = false
                }
            } catch {
                await MainActor.run {
                    self.deployErrorMessage = "Deployment failed: \(error.localizedDescription)"
                    self.isDeploying = false
                    appState.syncActivityLog(from: bridgeWrapper.bridge)
                }
            }
        }
    }
}
